### PR TITLE
AOT for reference types and bulk memory operations

### PR DESCRIPTION
**Pull request by ibmibmibm**

### CODE REVIEW

**Comment by q82419** [comment]
This may not be WASM 1.1.
The reference types and bulk memory operations proposals are not merged into WASM official spec yet.
I think it will be WASM 1.2 or others in the future.

**Comment by ibmibmibm** [comment]
> 
> 
> This may not be WASM 1.1.
> The reference types and bulk memory operations proposals are not merged into WASM official spec yet.
> I think it will be WASM 1.2 or others in the future.

Ok, I'll change commit messages

File: lib/interpreter/instantiate/function.cpp [hunk 1]
```diff
@@ -23,6 +23,11 @@ Expect<void> Interpreter::instantiate(
     auto NewFuncInst = std::make_unique<Runtime::Instance::FunctionInstance>(
         ModInst.Addr, *FuncType, CodeSegs[I]->getLocals(),
         CodeSegs[I]->getInstrs());
+    NewFuncInst->setSymbol(CodeSegs[I]->getSymbol());
```
Comment (hydai) [inline lib/interpreter/instantiate/function.cpp:26]: I have no idea about this line and the following lines.
It looks like 26 & (28-30) are doing the same stuff but the above one doesn't check the Symbol before calling the `setSymbol` function

File: lib/aot/compiler.cpp [hunk 2]
```diff
@@ -163,12 +159,7 @@ struct SSVM::AOT::Compiler::CompileContext {
         Trap(llvm::Function::Create(
             llvm::FunctionType::get(VoidTy, {Int32Ty}, false),
             llvm::Function::PrivateLinkage, "trap", LLModule)),
-        Likely(llvm::MDTuple::getDistinct(
-            LLContext, {llvm::MDString::get(LLContext, "branch_weights"),
-                        llvm::ConstantAsMetadata::get(llvm::ConstantInt::get(
-                            LLContext, llvm::APInt(32, 2000))),
-                        llvm::ConstantAsMetadata::get(llvm::ConstantInt::get(
-                            LLContext, llvm::APInt(32, 0)))})) {
+        Likely(llvm::MDBuilder(LLContext).createBranchWeights(2000, 0)) {
```
Comment (hydai) [inline lib/aot/compiler.cpp:162]: Why we need to set 2000 here? Is this a magic number?

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/aot/compiler.cpp (hunk 2 above)
Comment (ibmibmibm) [reply lib/aot/compiler.cpp:162]: It's a magic number at https://github.com/llvm/llvm-project/blob/991e86156c88dfa64c908fe36da3a1945c3d2e50/llvm/lib/Transforms/Scalar/LowerExpectIntrinsic.cpp#L50
Looks like llvm have expect Intrinsic now, I'll change this to an intrinsic call.

File: lib/interpreter/instantiate/function.cpp (hunk 1 above)
Comment (ibmibmibm) [reply lib/interpreter/instantiate/function.cpp]: Looks like I forgot to remove this line.

**Review by hydai — APPROVED** [review]
LGTM

**Pull request merged**